### PR TITLE
Add blake2b-256 algorithm

### PR DESCRIPTION
**Pull request by jku**
Fixes: #282 

### Description of the changes being introduced by the pull request:

Add 'blake2b-256' as a new supported algorithm: it is really the same algorithm as blake2b, just half the digest size.

Refactors hash tests a little to ensure that blake2* gets tested but that they do not get tested on pyca_crypto (which does not support them).

I've manually verified that we now get same results as Warehouse :)

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: tests/test_hash.py [hunk 1]
```diff
@@ -45,13 +45,89 @@
 
 class TestHash(unittest.TestCase):
 
-  def _run_with_all_hash_libraries(self, test_func):
-    for lib in securesystemslib.hash.SUPPORTED_LIBRARIES:
+  @staticmethod
+  def _get_algorithms(library):
+    algorithms = ['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+    if library != 'pyca_crypto':
```
Comment (trishankatdatadog) [inline tests/test_hash.py:51]: I'm not sure if this should be so dynamic between every function call. Also, it assumes that the library has been/not been loaded/unloaded.

**Review by trishankatdatadog — APPROVED** [review]
LGTM, I can't think of weaknesses. Thanks!

**Comment by jku** [comment]
Huh, blake2* is only supported in hashlib from python3.6 apparently

**Comment by jku** [comment]
I think test rig looks ok now: it should test every algorithm if the library and python version allow.
What I don't like is that it's not very easy to see that he correct algorithms did get tested -- but I think it's better than before.

**Review by joshuagl — COMMENTED** [review]
Should we be throwing some kind of error if a caller is trying to generate blake2* digests with Python < 3.6 or pyca?

**Comment by jku** [comment]
It does throw an error... and yes: if the tests checked for that, they would be much easier to reason about

**Comment by jku** [comment]
I plan to squash the commits into two: but did not do it yet in case Joshua wants to see diff vs the reviewed version.

Notes for reviewers:
* ended up refactoring the `_do_{algorithm}_update()` functions into a single function
* the supported algorithms for a test are still not listed in every test but now it's pretty easy to reason and verify what is expected to fail and what is expected to succeed: `_run_with_all_hash_libraries()` decides that and can easily be made to print out those decisions if they need to be checked
* found an inconsistency in my new code: see the TypeError handling in hash.py (yay for testing)

**Review by joshuagl — APPROVED** [review]
This looks much better, thanks @jku! 

I wonder if it would be clearer or not if we were using the [`unittest.skip*` decorators](https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures) to explicitly skip test cases where we don't expect them to work. That might lead to fewer unexpected errors than only accepting failures if it's an `UnsupportedAlgorithmError`, though does require explicit tests for the `UnsupportedAlgorithmError` on configurations which don't support it.

Let me know what you think.

**Comment by jku** [comment]
> I wonder if it would be clearer or not if we were using the [`unittest.skip*` decorators](https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures) to explicitly skip test cases where we don't expect them to work.

I like the fact that we now test every combination and actually verify the correct Error so skip doesn't sound right. A `@ExpectedFailureIf(condition, error, reason)` might be useful but that does not exist out of the box.

Even with that, I don't think we can make it work since the tests are not "unique": a single test case tests all libraries so we can't condition the skip/expectedfailure it on the library argument. Parameterizing 'library' (and maybe 'algorithm') would enable that but that means an additional test dependency on something like https://github.com/wolever/parameterized -- I think that's too much.

TL;DR: we should keep parameterizing in mind as an option in future, but for now the decorations are IMO not helpful

**Comment by joshuagl** [comment]
Thanks for taking the time to reason through that and discuss with me! Merging as is with a plan to look at parameterising tests in future.
